src/metabase/lib/schema/filter.cljc

```diff
@@ -3,15 +3,116 @@
+   [metabase.lib.schema.ref :as ref]
    [metabase.util.malli.registry :as mr]))
 
-(mr/def ::=
+(def ^:private eq-comparable
+  ;; avoid circular refs between these namespaces.
+  [:schema [:ref :metabase.lib.schema.expression/equality-comparable]])
+
+(def ^:private orderable
+  ;; avoid circular refs between these namespaces.
+  [:schema [:ref :metabase.lib.schema.expression/orderable]])
+
+(defn- defclause
+  [sch & args]
+  {:pre [(qualified-keyword? sch)]}
+  (mr/def sch
+    [:catn
+     [:clause [:= (keyword (name sch))]]
+     [:options ::common/options]
+     (into [:args] args)]))
+
+(doseq [op [::and ::or]]
+  (defclause op [:repeat {:min 2} [:schema [:ref ::filter]]]))
+
+(defclause ::not
+  [:schema [:ref ::filter]])
+
+(doseq [op [::= ::!=]]
+  (defclause op
+    [:repeat {:min 2} eq-comparable]))
+
+(doseq [op [::< ::<= ::> ::>=]]
+  (defclause op
+    [:cat orderable orderable]))
+
+(defclause ::between
+  [:catn [:field orderable] [:min orderable] [:max orderable]])
+
+;; sugar: a pair of `:between` clauses
+(defclause ::inside
+  [:catn
+   [:lat-field orderable]
+   [:lon-field orderable]
+   [:lat-max   orderable]
+   [:lon-min   orderable]
+   [:lat-min   orderable]
+   [:lon-max   orderable]])
+
+;; [:= ... nil], [:!= ... nil], [:or [:= ... nil] [:= ... ""]], [:and [:!= ... nil] [:!= ... ""]]
+(doseq [op [::is-null ::not-null ::is-empty ::not-empty]]
+  (defclause op
+    ::ref/field))
+
+(def ^:private string-filter-options
+  [:map [:case-sensitive {:optional true} :boolean]]) ; default true
+
+(def ^:private string
+  [:schema [:ref :metabase.lib.schema.expression/string]])
+
+;; [:does-not-contain ...] = [:not [:contains ...]]
+(doseq [op [::starts-with ::ends-with ::contains ::does-not-contain]]
+  (mr/def op
+    [:catn
+     [:clause [:= (keyword (name op))]]
+     [:options [:merge ::common/options string-filter-options]]
+     [:args [:catn [:whole string] [:part string]]]]))
+
+(def ^:private time-interval-options
+  [:map [:include-current {:optional true} :boolean]]) ; default false
+
+(def ^:private relative-datetime-unit
+  [:enum :default :minute :hour :day :week :month :quarter :year])
+
+;; SUGAR: rewritten as a filter clause with a relative-datetime value
+(mr/def ::time-interval
   [:catn
-   [:clause [:= :=]]
-   [:options ::common/options]
-   ;; avoid circular refs between these namespaces.
-   [:args [:+ {:min 2} :metabase.lib.schema.expression/equality-comparable]]])
+   [:clause [:= :time-interval]]
+   [:options [:merge ::common/options time-interval-options]]
+   [:args [:catn
+           [:field ::ref/field]
+           [:n [:or :int [:enum :current :last :next]]]
+           [:unit relative-datetime-unit]]]])
+
+(defclause ::segment
+  [:catn [:segment-id [:or ::common/int-greater-than-zero ::common/non-blank-string]]])
+
+(defclause ::case
+  [:+ [:catn
+       [:pred [:schema [:ref ::filter]]]
+       [:expr [:schema [:ref :metabase.lib.schema.expression/boolean]]]]])
```

Comment:
`:case` expression with boolean results. I wonder if we'll run into difficulties when `:case` expressions for other types are added.